### PR TITLE
mon/OSDMonitor: fix integer underflow of check_pg_num

### PR DESCRIPTION
**Pull request by 5cs**
Underflow of the `uint64_t projected` variable occurs when
the sum of current acting pg num and new pg num we specified
is less than the pg num calculated from pg info.

Signed-off-by: Tongliang Deng <dengtongliang@gmail.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by 5cs** [comment]
it follows c726ce9e5088b30d29e0db5c0ecc8c03fe41da1d

**Comment by 5cs** [comment]
@tchaikov could you please help review this pr, thx.

**Comment by 5cs** [comment]
jenkins test make check

File: src/osd/OSDMapMapping.h [hunk 1]
```diff
@@ -247,6 +247,15 @@ class OSDMapMapping {
 	row[4 + size + i] = up[i];
       }
     }
+
+    uint64_t get_num_acting_pgs() {
+      uint64_t num_acting_pgs = 0;
+      for (size_t ps = 0; ps < pg_num; ++ps) {
+        const int32_t *row = &table[row_size() * ps];
```
Comment (tchaikov) [inline src/osd/OSDMapMapping.h:254]: could extract `row_size()` out, and cache it.

File: src/osd/OSDMapMapping.h [hunk 2]
```diff
@@ -247,6 +247,15 @@ class OSDMapMapping {
 	row[4 + size + i] = up[i];
       }
     }
+
+    uint64_t get_num_acting_pgs() {
```
Comment (tchaikov) [inline src/osd/OSDMapMapping.h:251]: might want to mark this method `const`.

File: src/osd/OSDMapMapping.h [hunk 3]
```diff
@@ -330,6 +339,12 @@ class OSDMapMapping {
     return acting_rmap[osd];
   }
 
+  uint64_t get_num_acting_pgs(int64_t pool) {
```
Comment (tchaikov) [inline src/osd/OSDMapMapping.h:342]: might want to mark this method `const`.

File: src/osd/OSDMapMapping.h [hunk 4]
```diff
@@ -330,6 +339,12 @@ class OSDMapMapping {
...
+    auto p = pools.find(pool);
+    ceph_assert(p != pools.end());
```
Comment (tchaikov) [inline src/osd/OSDMapMapping.h:344]: i'd prefer using `assert()` instead. but i am fine either way.

File: src/osd/OSDMapMapping.h (hunk 1 above)
Comment (5cs) [reply src/osd/OSDMapMapping.h]: fixed

File: src/osd/OSDMapMapping.h (hunk 2 above)
Comment (5cs) [reply src/osd/OSDMapMapping.h]: fixed

File: src/osd/OSDMapMapping.h (hunk 3 above)
Comment (5cs) [reply src/osd/OSDMapMapping.h]: fixed

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test api

**Comment by ljflores** [comment]
jenkins test api

**Comment by ljflores** [comment]
http://pulpito.front.sepia.ceph.com/yuriw-2022-01-15_05:47:18-rados-wip-yuri8-testing-2022-01-14-1551-distro-default-smithi/
http://pulpito.front.sepia.ceph.com/yuriw-2022-01-17_17:14:22-rados-wip-yuri8-testing-2022-01-14-1551-distro-default-smithi/

Failures, unrelated:
    https://tracker.ceph.com/issues/45721
    https://tracker.ceph.com/issues/50280
    https://tracker.ceph.com/issues/53827
    https://tracker.ceph.com/issues/51076
    https://tracker.ceph.com/issues/53807
    https://tracker.ceph.com/issues/53842

Details:
    Bug_#45721: CommandFailedError: Command failed (workunit test rados/test_python.sh) FAIL: test_rados.TestWatchNotify.test
    Bug_#50280: cephadm: RuntimeError: uid/gid not found - Ceph
    Bug_#53827: cephadm exited with error code when creating osd. - Ceph - Orchestrator
    Bug_#51076: "wait_for_recovery: failed before timeout expired" during thrashosd test with EC backend.
    Bug_#53807: Dead jobs in rados/cephadm/smoke-roleless{...} - Ceph - Orchestrator
    Bug_#53842: cephadm/mds_upgrade_sequence: KeyError: 'en***'

**Pull request merged**